### PR TITLE
Offer methods for underlying values

### PR DESCRIPTION
**Pull request by TheConstructor**
While often `int` may be a good enough choice, especially for projects with enums based on different types it can become hard to track, which underlying type belongs to which enum. An added benefit is, that (as you don't support `"D"`-format-string) you can use `AsUnderlyingType` to have a numeric output, if needed. Also the build in `GetValuesAsUnderlyingType` is of course a lot slower than the quite trivial generated alternative.

### CODE REVIEW

File: tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt [hunk 1]
```diff
@@ -470,6 +470,26 @@ namespace MyTestNameSpace
...
+        /// <summary>
+        /// Cast a value of the underlying type (<c>int</c>) to <see cref="global::MyTestNameSpace.MyEnum" />.
+        /// This is mainly a convenience method.
+        /// </summary>
+        /// <returns>The value of the underlying type cast to <see cref="global::MyTestNameSpace.MyEnum" />.</returns>
+        public static global::MyTestNameSpace.MyEnum FromUnderlyingType(int value)
+        {
+            return (global::MyTestNameSpace.MyEnum) value;
+        }
+
```
Comment (andrewlock) [inline tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt:492]: I can see the value of `AsUnderlyingType` when you don't know the underlying type, but I'm not sure I see the value in this one? 🤔 Did you add it primarily for "symmetry" or is there a scenario that you consider this to be useful?

**Review by andrewlock — COMMENTED** [review]
Thanks for this! I think it looks good! 😄 I only have a couple of questions/requests:
- Could we add an integration test for non-`int` enums (e.g. `byte`)
- Could we add an integration test for enums with duplicate underlying values/combinations
- Do we need the `FromUnderlyingType` enum?

Thanks 🙂

File: tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt (hunk 1 above)
Comment (TheConstructor) [reply tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt:492]: I was thinking about DB/DTO interaction, were this can safe you from writing an explicit cast, if you use the appropriate numeric type; I briefly thought about making this an extension on the underlying type, but I thought that would clutter `int` and maybe other too much. And yes, I like the symmetry, too.

Comment (andrewlock) [reply tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt:492]: > I briefly thought about making this an extension on the underlying type, but I thought that would clutter `int`

💯 agree on this one, we definitely shouldn't do that


> I was thinking about DB/DTO interaction, were this can safe you from writing an explicit cast, if you use the appropriate numeric type;

I guess what I'm pushing back on is that this is any _better_ than an explicit cast 🤔 I guess I still can't see that it improves the code you write. IMO an explicit cast is just better and clearer 🤔 I can't think of a scenario off the top of my head where the helper is required to avoid complexity - could you provide an example?

FWIW I don't really feel like the "symmetry" argument makes is worth it - the source generator pretty much generates "constant" code that doesn't need a source generator, and could be written as a single generic helper for _all_ enums 🤷

**Comment by TheConstructor** [comment]
Oh, and I have a question for you: `ToStringFast()` always contains `_ => value.ToString(),` as default-branch. What do you think about `_ => value.AsUnderlyingType().ToString(),` for non-flag enums?

For flag-enums falling back to the build-in `enum.ToString()` is probably a good idea, as it offers the currently not supported comma-separated list of flags e.g. `(F.One | F.Two).ToString()` yields `One, Two`.

File: tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt (hunk 1 above)
Comment (TheConstructor) [reply tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt]: I still feel like it would be nice to have, but the other stuff is definitely useful more often. I pushed a commit without `FromUnderlyingType`

Comment (andrewlock) [reply tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsInChildNamespace.verified.txt]: Cool, thank you. We can always add it in later if it _does_ have value, but the fact you can easily create a generic version that works with _any_ enum means I don't expect it will have a lot of value

**Comment by andrewlock** [comment]
> There already been tests around LongEnum,

Cool, I forgot about that and didn't spot them in the PR, so sounds good 🙂 

> I added DuplicateValueByteEnum, that fulfils both aspects. Interestingly enough, duplicate underlying values mostly caused problems in the already existing extension-methods.

Ah, yeah, [this is a know issue](https://github.com/andrewlock/NetEscapades.EnumGenerators/issues/66) actually, I forgot about that. As mentioned in that PR, I'm not sure whether we _want_ to handle in general, and I don't think we should handle it in this PR either way 🤔 Sorry for the side-quest there, I knew there was _something_ at issue 😅 

> Adding this enum also surfaced a behavioural difference in enum.ToString() between .NET Framework 4.8, .NET <= 6 and .NET >= 7. I somehow assumed, that it would have always used the first name for a given value, but it seems it only does so starting with .NET 7. 

Could you explain this? 🤔 Is this _only_ related to duplicate case labels?

> I am not sure, that there is any value in being a 100% replacement for pre .NET 7 ToString()

IMO there absolutely _is_ value in that. I, for one, explicitly need it 😄 Is this _only_ an issue with the duplicate case labels scenario? If so, I'd probably personally rather we explicitly didn't support that scenario at all (as is the case prior to this PR)

> Oh, and I have a question for you: `ToStringFast()` always contains `_ => value.ToString(),` as default-branch. What do you think about `_ => value.AsUnderlyingType().ToString(),` for non-flag enums?
> 
> For flag-enums falling back to the build-in `enum.ToString()` is probably a good idea, as it offers the currently not supported comma-separated list of flags e.g. `(F.One | F.Two).ToString()` yields `One, Two`.

I'm torn 🤔 On the one hand, I _like_ the fact that we fallback to the original approach for anything outside the "standard" values. It's like a safety-net 😄 But I agree, there's not a good reason _not_ to do your suggestion AFAICT 🤔 I'd be inclined to do it in a separate PR though 🙂

**Comment by TheConstructor** [comment]
> > I added DuplicateValueByteEnum, that fulfils both aspects. Interestingly enough, duplicate underlying values mostly caused problems in the already existing extension-methods.
> 
> Ah, yeah, [this is a know issue](https://github.com/andrewlock/NetEscapades.EnumGenerators/issues/66) actually, I forgot about that. As mentioned in that PR, I'm not sure whether we _want_ to handle in general, and I don't think we should handle it in this PR either way 🤔 Sorry for the side-quest there, I knew there was _something_ at issue 😅
> 
> > Adding this enum also surfaced a behavioural difference in enum.ToString() between .NET Framework 4.8, .NET <= 6 and .NET >= 7. I somehow assumed, that it would have always used the first name for a given value, but it seems it only does so starting with .NET 7.
> 
> Could you explain this? 🤔 Is this _only_ related to duplicate case labels?
> 
> > I am not sure, that there is any value in being a 100% replacement for pre .NET 7 ToString()
> 
> IMO there absolutely _is_ value in that. I, for one, explicitly need it 😄 Is this _only_ an issue with the duplicate case labels scenario? If so, I'd probably personally rather we explicitly didn't support that scenario at all (as is the case prior to this PR)

Yes, it only relates to cases where two names are given to the same underlying value. Given
```cs
    [EnumExtensions]
    public enum DuplicateValueByteEnum : byte
    {
        First = 0,
        Min = First,
        Second = 1,
        [SuppressMessage("Design", "CA1069:Enums should not have duplicate values")]
        Max = 1,
    }
```
it seems that all my local .NETs (4.8, 2.1, 3.1, 6, 7, 8, 9) agree, that `((DuplicateValueByteEnum)1).ToString()` is `Second`, but `((DuplicateValueByteEnum)0).ToString()` is `Min` according to .NET 4.8, 2.1, 3.1 and 6, but `First` according to .NET 7, 8 and 9.

I prefer `First` and would like to argue, that always picking the first definition of a value is a behaviour you can easily plan with. I am not sure, if the older versions take the name and use it for ordering purposes, or if there is some deterministic randomness like a hash being involved.

> > Oh, and I have a question for you: `ToStringFast()` always contains `_ => value.ToString(),` as default-branch. What do you think about `_ => value.AsUnderlyingType().ToString(),` for non-flag enums?
> > For flag-enums falling back to the build-in `enum.ToString()` is probably a good idea, as it offers the currently not supported comma-separated list of flags e.g. `(F.One | F.Two).ToString()` yields `One, Two`.
> 
> I'm torn 🤔 On the one hand, I _like_ the fact that we fallback to the original approach for anything outside the "standard" values. It's like a safety-net 😄 But I agree, there's not a good reason _not_ to do your suggestion AFAICT 🤔 I'd be inclined to do it in a separate PR though 🙂

I can do that in a second PR, if you prefer. It's just if you go "we are faster than default-.NET", you should probably not just go through all of it again. The only case for non-flags, were there is a benefit, is the case, were source-generation no longer works.

**Comment by andrewlock** [comment]
> I prefer First and would like to argue, that always picking the first definition of a value is a behaviour you can easily plan with. I am not sure, if the older versions take the name and use it for ordering purposes, or if there is some deterministic randomness like a hash being involved.
> 
> Not supporting the case at all is maybe also not ideal. There could e.g. be a warning. In the end everything besides ToString is well defined

I'm very undecided about this tbh, but either way, I _don't_ think we should address it in this PR. The indeterministic behaviour between versions is as good a reason as any to _not_ support these IMO. Either way, I think would should remove the handling of those enums in this PR to unblock this and get your feature merged 😄 

> Not supporting the case at all is maybe also not ideal. There could e.g. be a warning

Yeah, that is where my gut leans, but apparently the .NET team are looking to _not_ support emitting diagnostics so going to see how that plays out 😅 

> I can do that in a second PR, if you prefer. It's just if you go "we are faster than default-.NET", you should probably not just go through all of it again. The only case for non-flags, were there is a benefit, is the case, were source-generation no longer works.

Thanks, I think a separate PR is best (I'm a fan of small units 🙂). From my PoV, I'm not really saying "we are faster than default-.NET" outside of the "happy" path. IMO, the fallback case is firmly in the edge case, so I'm _much_ less concerned with optimizing that. That said, I think your suggestion still makes sense, just don't want to block this PR any further!

**Comment by TheConstructor** [comment]
> > Not supporting the case at all is maybe also not ideal. There could e.g. be a warning
> 
> Yeah, that is where my gut leans, but apparently the .NET team are looking to _not_ support emitting diagnostics so going to see how that plays out 😅

You/we/whoever could bundle an analyzer, I guess, but I think it's a case, that is rare enough, to be solved by documentation. Also it might make your library more predictable, if you target multiple .NET-versions.

Another approach would be a parameter on the annotation, that you accept this deviation from `Enum.ToString`

> > I can do that in a second PR, if you prefer. It's just if you go "we are faster than default-.NET", you should probably not just go through all of it again. The only case for non-flags, were there is a benefit, is the case, were source-generation no longer works.
> 
> Thanks, I think a separate PR is best (I'm a fan of small units 🙂). From my PoV, I'm not really saying "we are faster than default-.NET" outside of the "happy" path. IMO, the fallback case is firmly in the edge case, so I'm _much_ less concerned with optimizing that. That said, I think your suggestion still makes sense, just don't want to block this PR any further!

I stripped out everything but adding `AsUnderlyingType(this)` and `GetValuesAsUnderlyingType()` from this PR

**Comment by andrewlock** [comment]
Thanks for this, it looks good to me! (There's an issue with the build on forks currently, hence the failed build)
